### PR TITLE
Gvcf from hvcf

### PR DESCRIPTION
**Pull request by lynnjo**
## Description

This branch takes a h.vcf file created from PHG pathing code and creates a gvcf file.  Indels are not resized in this version, and it currently is not implemented with co-routines.  The intention is to add that later.

## Type of change

_What type of changes does your code introduce? Put an `x` in boxes that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated relevant documentation

## Changelog entry

_Please add a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Initial code to create a gvcf file from a PHG pathing created hvcf file.

### CODE REVIEW

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+        // read and store ref file data for later use when creating the vcf sequence dictionary
+        val refSeq = CreateMafVcf().buildRefGenomeSeq(referenceFile)
+        buildGvcfFromHvcf(dbPath, refSeq, outputDir, hvcfDir, condaEnvPrefix)
+    }
+
+    fun buildGvcfFromHvcf(dbPath: String, refSeq: Map<String, NucSeq>, outputDir: String, hvcfDir: String, condaEnvPrefix: String) {
+        // get list of hvcf files
+        // walk the gvcf directory process files with g.vcf.gz extension
+        File(hvcfDir).walk().filter { !it.isHidden && !it.isDirectory }
+            .filter { it.name.endsWith("h.vcf.gz")  || it.name.endsWith("h.vcf")  }.toList()
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:96]: Does this .toList() need to be there?

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+                val records = processSingleHVCF(refSeq, outputDir, hvcfFile, dbPath,condaEnvPrefix)
+
+                val sample = hvcfFile.toString().substringAfterLast("/").substringBefore(".")
+                val gvcfFile = "$outputDir/${sample}.g.vcf"
+                exportVariantContext(sample,records,gvcfFile, refSeq,setOf())
+            }
+
+    }
+
+    fun processSingleHVCF(refSeq:Map<String,NucSeq>, outputDir:String, hvcfFile: File, dbPath: String, condaEnvPrefix: String): List<VariantContext> {
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:108]: Can this have a more specific name?  Maybe processSingleHVCFIntoVariantContext or something a bit simpler?

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+        // altHeaders map with the key of the current record's ID.
+        val sampleToRefRanges = mutableMapOf<String, MutableList<ReferenceRange>>()
+
+        // Get the sampleName from the hvcf file.  This is the name
+        // without the path, and without the extension of h.vcf, hvcf, h.vcf.gz or hvcf.gz
+        val pathSample = hvcfFile.toString().substringAfterLast("/").substringBefore(".")
+
+        var rangesSkipped = 0
+        // process the hvcf records into the sampleToRefRanges map
+        reader.forEach { context ->
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:132]: Consider using a map{} so you can reduce the mutability of sampleToRefRanges.

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+                    // meaning the variant may overlap the region.  We need to adjust the asm positions
+                    val fixedVariants = fixPositions(Pair(Position(regionChrom,regionStart),Position(regionChrom,regionEnd)), listOf(currentVariant))
+                    // Add the fixedVariants to the refRangeToVariantContext map, for the current range
+                    val currentVariants = refRangeToVariantContext.getOrPut(range, { mutableListOf() })
+                    currentVariants.addAll(fixedVariants)
+                    refRangeToVariantContext[range] = currentVariants
+                    tempVariants.clear()
+                    break
+                }
+                if(CreateMafVcf().variantFullyContained(Pair(Position(regionChrom,regionStart),Position(regionChrom,regionEnd)), currentVariant)) {
```
Comment (zrm22) [inline src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:369]: We should probably move these CreateMafVCF() methods to a utility function.

File: src/test/kotlin/net/maizegenetics/phgv2/cli/Hvcf2GvcfTest.kt [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        File(lineBHvcf).copyTo(File(lineBPathHvcf))
+
+        // Run hvcf2gvcf on the copied file
+        val hvcf2gvcf = Hvcf2Gvcf()
+        result = hvcf2gvcf.test("--db-path ${dbPath} --hvcf-dir $testGVCFdir --output-dir ${testGVCFdir} --reference-file ${refFasta}")
+
+        // TODO
+        // Need some assertions here
+        println("done !!")
+
```
Comment (zrm22) [inline src/test/kotlin/net/maizegenetics/phgv2/cli/Hvcf2GvcfTest.kt:118]: missing assertions

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt (hunk 1 above)
Comment (lynnjo) [reply src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:96]: nope!  I removed it.

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt (hunk 2 above)
Comment (lynnjo) [reply src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:108]: I changed to processHVCFtoVariantContext

File: src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt (hunk 3 above)
Comment (lynnjo) [reply src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt:132]: recoded with map{}

File: src/test/kotlin/net/maizegenetics/phgv2/cli/Hvcf2GvcfTest.kt (hunk 5 above)
Comment (lynnjo) [reply src/test/kotlin/net/maizegenetics/phgv2/cli/Hvcf2GvcfTest.kt:118]: yup - it's in the TODO.   - working on it.

**Review by zrm22 — APPROVED** [review]

**Review by tcasstevens — APPROVED** [review]

File: docs/imputation.md [hunk 6]
```diff
@@ -671,3 +679,24 @@ This command takes three parameters:
 * `--db-path` - Path to the directory containing the TileDB instances.
 * `--threads` - Number of threads for use by the TileDB loading
   procedure.
+
+### Create g.vcf files
```
Comment (btmonier) [inline docs/imputation.md:683]: Add "(optional)" after title for extra clarity

Comment (lynnjo) [reply docs/imputation.md]: done

**Review by aberthel — APPROVED** [review]

**Review by btmonier — APPROVED** [review]

**Review by tcasstevens — APPROVED** [review]

**Pull request merged**